File: live_feed.py
"""
Vercel serverless endpoint: GET /api/live_feed?count=N

Generates simulated transactions and scores them with the same
dependency-free heuristic as api/predict.py, mirroring app.py's
random_tx() for the local demo.
"""
import json
import math
import random
from datetime import datetime
from http.server import BaseHTTPRequestHandler
from urllib.parse import parse_qs, urlparse

DISCOS = ["AEDC", "EKEDC", "IKEDC", "EEDC", "PHEDC", "JEDC", "BEDC", "KAEDCO", "KEDCO", "YEDC", "IBEDC"]
STATES = ["Abuja", "Lagos", "Rivers", "Kano", "Kaduna", "Enugu", "Imo", "Oyo", "Delta", "Edo", "Borno"]
CHANNELS = ["web", "mobile_app", "ussd", "agent", "pos"]


def clamp(v, lo=0.0, hi=1.0):
    return max(lo, min(hi, v))


def rule_based(amount, hour, velocity, is_new, is_vpn,
               sim_swap, num_meters, fail_logins, distance, account_age):
    night = 1.0 if (hour < 6 or hour >= 22) else 0.0
    base = (
        clamp(math.log1p(amount) / 13.0) * 0.24 +
        clamp(velocity / 20.0) * 0.16 +
        is_vpn * 0.16 +
        sim_swap * 0.14 +
        is_new * 0.10 +
        night * 0.08 +
        clamp(fail_logins / 8.0) * 0.06 +
        clamp(num_meters / 10.0) * 0.04 +
        (1.0 - clamp(account_age / 3650.0)) * 0.04 +
        clamp(distance / 500.0) * 0.02
    )
    dnn_p = clamp(base * 0.95 + 0.02)
    lstm_p = clamp(base * 0.90 + 0.04)
    ensemble = dnn_p * 0.6 + lstm_p * 0.4
    return dnn_p, lstm_p, ensemble


def predict(raw):
    amount = float(raw.get("amount", 0) or 0)
    hour = int(raw.get("hour", 0) or 0)
    velocity = int(raw.get("velocity", 0) or 0)
    is_new = int(raw.get("is_new_device", 0) or 0)
    is_vpn = int(raw.get("is_vpn", 0) or 0)
    sim_swap = int(raw.get("sim_swap", 0) or 0)
    num_meters = int(raw.get("num_meters", 0) or 0)
    fail_logins = int(raw.get("fail_logins", 0) or 0)
    distance = float(raw.get("distance", 0) or 0)
    account_age = int(raw.get("account_age", 0) or 0)

    dnn_p, lstm_p, ensemble = rule_based(
        amount, hour, velocity, is_new, is_vpn,
        sim_swap, num_meters, fail_logins, distance, account_age,
    )

    fraud = ensemble >= 0.5
    level = "HIGH" if ensemble > 0.7 else "MEDIUM" if ensemble > 0.4 else "LOW"
    color = "#E24B4A" if level == "HIGH" else "#BA7517" if level == "MEDIUM" else "#639922"
    return {
        "success": True,
        "fraud": bool(fraud),
        "probability": round(ensemble * 100, 2),
        "dnn_prob": round(dnn_p * 100, 2),
        "lstm_prob": round(lstm_p * 100, 2),
        "risk_level": level,
        "risk_color": color,
        "verdict": "FRAUDULENT" if fraud else "LEGITIMATE",
    }


def random_tx():
    fraud = random.random() < 0.09
    disco = random.choice(DISCOS)
    state = random.choice(STATES)
    channel = random.choice(CHANNELS)
    if fraud:
        amount = random.randint(50000, 300000)
        raw = dict(amount=amount, hour=random.choice(list(range(6)) + list(range(22, 24))),
                   velocity=random.randint(8, 25), is_new_device=random.choice([0, 1, 1]),
                   is_vpn=random.choice([0, 1, 1]), sim_swap=random.choice([0, 1]),
                   num_meters=random.randint(3, 12), fail_logins=random.randint(3, 8),
                   distance=round(random.uniform(50, 500), 1), account_age=random.randint(1, 60))
    else:
        amount = random.randint(500, 25000)
        raw = dict(amount=amount, hour=random.randint(7, 21),
                   velocity=random.randint(1, 3), is_new_device=random.choice([0, 0, 0, 1]),
                   is_vpn=random.choice([0, 0, 0, 1]), sim_swap=0,
                   num_meters=random.randint(1, 2), fail_logins=random.randint(0, 1),
                   distance=round(random.uniform(0, 15), 1), account_age=random.randint(100, 2000))
    result = predict(raw)
    return {
        "id": "TXN" + "".join(str(random.randint(0, 9)) for _ in range(7)),
        "disco": disco, "state": state, "channel": channel,
        "amount": f"NGN {amount:,}",
        "timestamp": datetime.now().strftime("%H:%M:%S"),
        **result,
    }


class handler(BaseHTTPRequestHandler):
    def _cors(self):
        self.send_header("Access-Control-Allow-Origin", "*")
        self.send_header("Access-Control-Allow-Methods", "GET, OPTIONS")
        self.send_header("Access-Control-Allow-Headers", "Content-Type")

    def do_OPTIONS(self):
        self.send_response(200)
        self._cors()
        self.end_headers()

    def do_GET(self):
        qs = parse_qs(urlparse(self.path).query)
        try:
            count = max(1, min(50, int(qs.get("count", ["5"])[0])))
        except ValueError:
            count = 5

        payload = json.dumps([random_tx() for _ in range(count)]).encode("utf-8")
        self.send_response(200)
        self.send_header("Content-Type", "application/json; charset=utf-8")
        self.send_header("Content-Length", str(len(payload)))
        self._cors()
        self.end_headers()
        self.wfile.write(payload)

File: live_feed.js
// Vercel serverless endpoint: GET /api/live_feed?count=N
//
// Generates simulated transactions and scores them with the same
// dependency-free heuristic as api/predict.js, mirroring app.py's
// random_tx() for the local demo.

const DISCOS = ["AEDC", "EKEDC", "IKEDC", "EEDC", "PHEDC", "JEDC", "BEDC", "KAEDCO", "KEDCO", "YEDC", "IBEDC"];
const STATES = ["Abuja", "Lagos", "Rivers", "Kano", "Kaduna", "Enugu", "Imo", "Oyo", "Delta", "Edo", "Borno"];
const CHANNELS = ["web", "mobile_app", "ussd", "agent", "pos"];

function clamp(v, lo = 0, hi = 1) {
  return Math.max(lo, Math.min(hi, v));
}

function ruleBased(amount, hour, velocity, isNew, isVpn, simSwap, numMeters, failLogins, distance, accountAge) {
  const night = hour < 6 || hour >= 22 ? 1 : 0;
  const base =
    clamp(Math.log1p(amount) / 13.0) * 0.24 +
    clamp(velocity / 20.0) * 0.16 +
    isVpn * 0.16 +
    simSwap * 0.14 +
    isNew * 0.1 +
    night * 0.08 +
    clamp(failLogins / 8.0) * 0.06 +
    clamp(numMeters / 10.0) * 0.04 +
    (1 - clamp(accountAge / 3650.0)) * 0.04 +
    clamp(distance / 500.0) * 0.02;
  const dnnP = clamp(base * 0.95 + 0.02);
  const lstmP = clamp(base * 0.9 + 0.04);
  const ensemble = dnnP * 0.6 + lstmP * 0.4;
  return { dnnP, lstmP, ensemble };
}

function predict(raw) {
  const { amount, hour, velocity, is_new_device, is_vpn, sim_swap, num_meters, fail_logins, distance, account_age } = raw;
  const { dnnP, lstmP, ensemble } = ruleBased(
    amount, hour, velocity, is_new_device, is_vpn, sim_swap, num_meters, fail_logins, distance, account_age
  );
  const fraud = ensemble >= 0.5;
  const level = ensemble > 0.7 ? "HIGH" : ensemble > 0.4 ? "MEDIUM" : "LOW";
  const color = level === "HIGH" ? "#E24B4A" : level === "MEDIUM" ? "#BA7517" : "#639922";
  return {
    success: true,
    fraud,
    probability: Math.round(ensemble * 10000) / 100,
    dnn_prob: Math.round(dnnP * 10000) / 100,
    lstm_prob: Math.round(lstmP * 10000) / 100,
    risk_level: level,
    risk_color: color,
    verdict: fraud ? "FRAUDULENT" : "LEGITIMATE",
  };
}

function randInt(a, b) {
  return Math.floor(Math.random() * (b - a + 1)) + a;
}
function choice(arr) {
  return arr[Math.floor(Math.random() * arr.length)];
}
function round1(v) {
  return Math.round(v * 10) / 10;
}

function randomTx() {
  const isFraud = Math.random() < 0.09;
  const disco = choice(DISCOS);
  const state = choice(STATES);
  const channel = choice(CHANNELS);
  let raw;
  let amount;

  if (isFraud) {
    amount = randInt(50000, 300000);
    raw = {
      amount,
      hour: choice([0, 1, 2, 3, 4, 5, 22, 23]),
      velocity: randInt(8, 25),
      is_new_device: choice([0, 1, 1]),
      is_vpn: choice([0, 1, 1]),
      sim_swap: choice([0, 1]),
      num_meters: randInt(3, 12),
      fail_logins: randInt(3, 8),
      distance: round1(Math.random() * (500 - 50) + 50),
      account_age: randInt(1, 60),
    };
  } else {
    amount = randInt(500, 25000);
    raw = {
      amount,
      hour: randInt(7, 21),
      velocity: randInt(1, 3),
      is_new_device: choice([0, 0, 0, 1]),
      is_vpn: choice([0, 0, 0, 1]),
      sim_swap: 0,
      num_meters: randInt(1, 2),
      fail_logins: randInt(0, 1),
      distance: round1(Math.random() * 15),
      account_age: randInt(100, 2000),
    };
  }

  const result = predict(raw);
  const id = "TXN" + Array.from({ length: 7 }, () => randInt(0, 9)).join("");
  const timestamp = new Date().toTimeString().slice(0, 8);

  return {
    id,
    disco,
    state,
    channel,
    amount: `NGN ${amount.toLocaleString("en-US")}`,
    timestamp,
    ...result,
  };
}

module.exports = (req, res) => {
  res.setHeader("Access-Control-Allow-Origin", "*");
  res.setHeader("Access-Control-Allow-Methods", "GET, OPTIONS");
  res.setHeader("Access-Control-Allow-Headers", "Content-Type");

  if (req.method === "OPTIONS") {
    res.status(200).end();
    return;
  }

  const rawCount = (req.query && req.query.count) || "5";
  let count = parseInt(Array.isArray(rawCount) ? rawCount[0] : rawCount, 10);
  if (!Number.isFinite(count)) count = 5;
  count = Math.max(1, Math.min(50, count));

  const txs = Array.from({ length: count }, randomTx);
  res.status(200).json(txs);
};
